### PR TITLE
SQL: check TRANSACTION_REGISTRY table schema on start [closes #309]

### LINKED ISSUE #309: TRT schema check
We would need some kind of schema check to ensure that the SQL-created table is in InnoDB format and has the expected schema.

It MUST check:
* field count and types
* table engine
* PRIMARY KEY is field[0] (i.e. `transaction_id`)

It SHOULD NOT check:
* field names
* indexes

Please, [examine](https://github.com/tempesta-tech/mariadb/issues/305#issuecomment-341088320) server start procedure to find optimal place for schema check.

Related to #305.

### PR DESCRIPTION
**Pull request by kevgs**
Function (which is a place to check) is called in non-bootstrap mode:
From `sql/mysqld.cc`
```c++
  if (!opt_bootstrap)
    servers_init(0);
```

Check exists in `servers_init()` because we need `THD` to open a table. And this is one of places where `THD` created.

### CODE REVIEW

File: sql/sql_servers.cc [hunk 1]
```diff
@@ -112,6 +112,54 @@ static void init_servers_cache_psi_keys(void)
...
+  if (!t->key_info || !t->key_info->key_part)
+    return false;
+
+  if (strcmp(t->key_info->key_part->field->field_name.str, "transaction_id"))
+    return false;
+
+  return true;
+}
+
+bool trt_table_correct(THD *thd, bool &correct)
```
Comment (midenok) [inline sql/sql_servers.cc]: Move to `TR_table::check()`. I don't see the reason why the code on opening the table should be duplicated.

File: sql/sql_servers.cc [hunk 2]
```diff
@@ -112,6 +112,54 @@ static void init_servers_cache_psi_keys(void)
 }
 #endif /* HAVE_PSI_INTERFACE */
 
+
+bool trt_table_type_is_correct(TABLE *t)
```
Comment (midenok) [inline sql/sql_servers.cc]: Don't see the reason why there is second function dedicated just for type checking.

File: sql/sql_servers.cc [hunk 3]
```diff
@@ -166,6 +215,18 @@ bool servers_init(bool dont_read_servers_table)
     DBUG_RETURN(TRUE);
   thd->thread_stack= (char*) &thd;
   thd->store_globals();
+
+  // Check transaction_registry table schema.
+  if (trt_table_correct(thd, trt_correct))
```
Comment (midenok) [inline sql/sql_servers.cc]: This has nothing to do with `mysql.servers` table. Place schema check into `mysqld_main()` after `Events::init()` (mysqld.cc:6024). Make your own temporary `thd`.

**Review by midenok — CHANGES_REQUESTED** [review]

**Comment by kevgs** [comment]
3 notes fixed.

File: sql/mysqld.cc [hunk 4]
```diff
@@ -6021,6 +6023,25 @@ int mysqld_main(int argc, char **argv)
...
+    THD *thd = new THD(0);
+    thd->thread_stack= (char*) &thd;
+    thd->store_globals();
+
+    {
+      TR_table trt(thd);
+      if (trt.check())
+      {
+        sql_print_error("%s schema is incorrect", trt.table_name);
+        exit(1);
```
Comment (midenok) [inline sql/mysqld.cc]: Use `unireg_abort(1)`. It is good to free `thd` here as well.

File: sql/table.cc [hunk 5]
```diff
@@ -8684,6 +8684,43 @@ bool TR_table::query_sees(bool &result, ulonglong trx_id1, ulonglong trx_id0,
...
+    return true;
+
+  if (table->field[2]->type() != MYSQL_TYPE_TIMESTAMP)
+    return true;
+
+  if (table->field[3]->type() != MYSQL_TYPE_TIMESTAMP)
+    return true;
+
+  if (table->field[4]->type() != MYSQL_TYPE_STRING)
+    return true;
```
Comment (midenok) [inline sql/table.cc]: Use `field_id_t`. Also check `FLD_ISO_LEVEL` for enum order if possible (see SHOW CREATE code).

**Pull request closed without merging**